crates/matrix-sdk/src/client/mod.rs

```diff
@@ -437,48 +455,115 @@ impl Client {
     /// }).await;
     /// # });
     /// ```
-    pub async fn register_event_handler<Ev, Ctx, H>(&self, handler: H) -> &Self
+    pub async fn register_event_handler<Ev, Ctx, H>(&self, handler: H) -> EventHandlerHandle
     where
         Ev: SyncEvent + DeserializeOwned + Send + 'static,
         H: EventHandler<Ev, Ctx>,
         <H::Future as Future>::Output: EventHandlerResult,
     {
         let event_type = H::ID.1;
-        self.inner.event_handlers.write().await.entry(H::ID).or_default().push(Box::new(
-            move |data| {
-                let maybe_fut = serde_json::from_str(data.raw.get())
-                    .map(|ev| handler.clone().handle_event(ev, data));
-
-                Box::pin(async move {
-                    match maybe_fut {
-                        Ok(Some(fut)) => {
-                            fut.await.print_error(event_type);
-                        }
-                        Ok(None) => {
-                            error!(
-                                "Event handler for {} has an invalid context argument",
-                                event_type
-                            );
-                        }
-                        Err(e) => {
-                            warn!(
-                                "Failed to deserialize `{}` event, skipping event handler.\n\
-                                 Deserialization error: {}",
-                                event_type, e,
-                            );
-                        }
+
+        let handler_function: Box<EventHandlerFn> = Box::new(move |data| {
+            let maybe_fut = serde_json::from_str(data.raw.get())
+                .map(|ev| handler.clone().handle_event(ev, data));
+
+            Box::pin(async move {
+                match maybe_fut {
+                    Ok(Some(fut)) => {
+                        fut.await.print_error(event_type);
                     }
-                })
-            },
-        ));
+                    Ok(None) => {
+                        error!("Event handler for {} has an invalid context argument", event_type);
+                    }
+                    Err(e) => {
+                        warn!(
+                            "Failed to deserialize `{}` event, skipping event handler.\n\
+                                Deserialization error: {}",
+                            event_type, e,
+                        );
+                    }
+                }
+            })
+        });
 
-        self
+        let handle_id = self.inner.event_handler_counter.fetch_add(1, SeqCst);
+
+        let handle = EventHandlerHandle { id: H::ID, addr: handle_id };
+
+        self.inner
+            .event_handlers
+            .write()
+            .await
+            .entry(H::ID)
+            .or_default()
+            .push(EventHandlerWrapper { handler_function, handle: handle.clone() });
```

Comment:
I don't see why we should save the "full" handle here, just the `handle_id` should be sufficient.

Reply:
I need the `EventHandlerHandle` each time the handler is called in `fn handle_sync_events_wrapped_with` to pass it to `EventHandlerData`. I could of course create a new instance of `EventHandlerHandle` each time the handler is called, but that is less efficient in my eyes.

Reply:
I'm pretty sure it has almost no cost. I would prefer if it was done that way, but it might also be harder. If it isn't easy / you don't want to do it, I'm okay merging this PR without that and revisiting it later.

Reply:
I just tried to change it, but to create a new `EventHandlerHandle` within `fn handle_sync_events_wrapped_with` I need the variable `ev_type` which, since it is inferred from `raw_event`, has a lifetime of `'a` instead of `'static'. 

I will leave it as is since that is alright with you.